### DISCUSSION TITLE
Can't debug wasm project after update to VS 17.8.6

### DISCUSSION THREAD

**Discussion by appel1** [AvaloniaUI/Avalonia#14570] [General]
## Describe the bug
When attempting to debug a browser showing about:blank appears and then goes away after a while and an error is shown in VS.

Seems like a VS issue but thought I should mention it here in case there's a known solution or workaround.

See https://developercommunity.visualstudio.com/t/Debug-of-Avalonia-WebAssembly-app-fails-/10576821

## To Reproduce
Create a new avalonia xplay solution using sorbet new avalonia.xplat
Debug browser peojwct

## Expected behavior
Successful debug as with 17.8.4

## Environment 

- OS: Windows 11
- Avalonia-Version: 11.0.7
- Same with both Edges and Chrome

**Comment by migubis** [comment]
We have observed a similar problem and found out that it is (in our case) connected with the usage of **UseUrls**. I have submitted it to Microsoft under 
https://developercommunity.visualstudio.com/t/Debugging-does-not-work-when-UseUrls-is/10582270

My problem is not connected to Avalonia in any way, but I thought maybe it helps to share my findings.

**Comment by maxkatz6** [comment]
I will convert this issue to discussion, as it has a very chance of being caused by Avalonia. 
But there still could be some workaround or useful links to share.

**Reply by gitasaurus** [reply]
I experience the same behavior, using Blazor SWA and the Azure Static Web Apps CLI.   In 17.8.5, it properly connects to the browser (in my case Microsoft Edge).  Once I upgrade to 17.8.6 it stops working providing the same **Failed to launch debug adapter** message as show in https://developercommunity.visualstudio.com/t/Debug-of-Avalonia-WebAssembly-app-fails-/10576821.  I am **not using Avalonia**.  Rolling back Visual Studio to 17.8.5 fixes the issue again.   I have tried this on a few different machines with two different SWAs and all exhibit the exact same behavior.

**Reply by anispider** [reply]
For a different data point, I experience this in 17.8.6 using the wizard to create a new xplat Avalonia project, the Browser project exhibits that same behavior as the OP. However if I create a "Blazor WebAssembly Standalone App" project it can launch and run in the Visual Studio debugger just fine in Edge. 

This is pretty much wild speculation on my part, but I have another issue which has cropped up with wasm Avalonia projects in the past month or two that used to work is that asynchronous HttpClient calls seem to be broken. Identical HttpClient code works as expected in a "Hello World" Blazor WASM app. In the Avalonia case, I can tell in the browser's dev tools that the http operation executes and returns valid data, but the "await client.GetAsync(...)" call never resumes. That is consistent across all the browsers I tried and whether or not I launched from VS or run a published build from a webserver. The only reason I bring that up here is that my wild speculation is that maybe both issues have a related root cause, in that perhaps something about the way the dotnet runtime is getting instanced and initialized for Avalonia no longer works. Comparing the startup JS for Avalonia vs Blazor is tough because a lot of the Blazor bootstrap JS  is minified typescript. The really odd duck is that why would anything in Avalonia affect the behavior of a core library? The thing that jumps to mind is something about how the hosting is setup. I've tried to ramp up logging, and I never see anything in the console output that would be indicative of an error. Lacking the ability to debug has made trying to track down the http issue more challenging. 

For what it's worth, I had no better luck with Rider when trying to debug, though the error message is at least a little more informative:

```
Error running '[My app]'
Blazor debug proxy error: redirect from /ws-proxy was expected, but the 404 status code was received. See logs for details.
```

**Reply by thevortexcloud** [reply]
> used to work is that asynchronous HttpClient calls seem to be broken

See: https://github.com/AvaloniaUI/Avalonia/issues/14022

**Reply by anispider** [reply]
@thevortexcloud OMG, THANK YOU! That fixed it. 

As far as the Debugging problem, I'll comment here if I find out anything else. It does feel like a Visual Studio issue. It's also worth nothing that I don't have much familiarity with Uno, but I setup a "Hello World" wasm project using it, and it had the same problem as with the Avalonia projects.

**Reply by thevortexcloud** [reply]
You are welcome! A lot of people have been having this problem recently. Most people also don't update their dotnet templates very often either so it's probably going to keep coming up for awhile. It also sounds like Uno possibly needs to update their templates as there was a breaking change (as mentioned in that issue) with .NET 8.

As for your debugging issue: Based on what I am reading Web assembly starts up some debugging server when instructed to do so. I am guessing that's not happening here for some reason.

https://learn.microsoft.com/en-us/aspnet/core/blazor/debug?view=aspnetcore-8.0&tabs=visual-studio

**Reply by gitasaurus** [reply]
FYI, I opened an issue in the **Azure/static-web-apps-cli** repo (Azure/static-web-apps-cli#801)

**Reply by anispider** [reply]
@thevortexcloud I think I had tried to update the inspectUri setting and also add the Microsoft.AspNetCore.Components.WebAssembly.DevServer reference, and it didn't make a difference. I can try again, looking for something else I might have overlooked. 

The debugging problem isn't technically blocking me, just slowing me down as I have to reply on console logging to debug. The async calls not working had me dead in the water though, you saved me a lot of time. Both the async call issue and the debugging issue felt like they cropped up around the same time, and since I happened to be using HttpClient as the initial async call that was failing, it seemed like maybe they could be related. I had assumed the async thing was a networking issue, and my google fu was failing me to find any relevant info.

**Reply by anispider** [reply]
For anyone following this, Visual Studio 17.9.0 came out today, and this issue is still present.

**Reply by mb-jp** [reply]
Any news on this? it's a big problem for me 😭

**Reply by maxkatz6** [reply]
You might try .NET 9 previews with preview Visual Studio.
I haven't tested, but if it was fixed in runtime, it should be available there.

**Reply by mb-jp** [reply]
I couldn't get it to work with .NET 9 previews and preview Visual Studio. 😟

**Reply by JW-CH** [reply]
Any update on this?

**Reply by DVaughan** [reply]
Today I updated VS to Version 17.10.0 Preview 7.0. The problem still persists. I realize this may be a VS issue, but would like to know if there is a work-around that someone is aware of to get the stock Avalonia web template to launch and attach to the browser. Some things I've tried that have not worked:
* Modified await dotnetRuntime.runMainAndExit to runMain
* Tried alternate URLs in launchSettings.json: "applicationUrl": "https://localhost:5001;http://localhost:5000",
* Changed csproj file TargetFramework from net7.0 to net8.0 (I had an issue with updating the webassembly workload unfortunately).

**Comment by yll690** [comment]
You can try the 360 browser: https://browser.360.cn/eex/index.html . My Edge and Chrome open the about:blank but the 360 browser can open https://localhost:5001/index.html

Or use IIS Express configuaration.

**Comment by JW-CH** [comment]
Still not resolved for me.

**Reply by anispider** [reply]
Same here. Luckily, my code runs on multiple platforms, so I use those to debug, otherwise this would be a show-stopper. It does seem like a VS bug, since it seemed to manifest immediately after the upgrade mentioned in the OP title, but then again brand new Blazor WASM projects debug just fine. 

If anyone on the team would like end user assistance on this, I'd love to help. Even if it is a VS bug, it seems to have a big impact on Avalonia development. I haven't tried Uno in a while, but back when this issue first appeared, Uno debugging was broken too.

**Reply by timunie** [reply]
@anispider we can offer this to users who have paid support. If you have a support agreement, reach out to the team.

**Reply by anispider** [reply]
@timunie I think you misunderstood me, I wasn't asking for help, I was offering it. This issue does impact me, but I can work around it in my scenario, as inconvenient as that may be. But whether or not the root cause lies with VS or Avalonia (I suspect it is VS), I would think the Avalonia team would have a lot of interest in getting it resolved, since it kind of takes away from one of Avalonia's key strengths, being able to share code between the browser and other platforms, and debugging code running in-browser. 

I really love the framework you have built, and want to see it grow in adoption, but if I knew nothing about it, and I was tasked in a corporate job with evaluating tech stacks for a new project, the broken debugging thing might lead me to blame the issue on Avalonia and conclude that it wasn't mature yet.

**Comment by Quantum404Void** [comment]
Can this problem be solved in VS? Because I've been having this issue recently as well

**Comment by anispider** [comment]
@WAASSTT I feel like this issue is pretty close to being solved. I'm not sure what has changed recently, but in VS 17.11.1 (the latest stable version as of the time of this comment), and with a brand new browser project using Avalonia 11.1.X I can at least launch in the debugger. Breakpoints don't seem to work, but it's a slight improvement from where this issue was a few months ago. My existing Avalonia projects still seem broken, so maybe there is some legacy cruft in my project files or launch settings that it doesn't like.

**Reply by anispider** [reply]
(And bringing my "legacy" projects up to Avalonia 11.1.3 didn't fix them, so whatever change allowed brand new projects to almost work, I think it was something in the Avalonia new project wizard templates)
